packages/firestore/exp/test/shim.ts

```diff
@@ -0,0 +1,768 @@
+  }
+}
+
+export class DocumentReference<T = legacy.DocumentData>
+  implements legacy.DocumentReference<T> {
+  constructor(readonly _delegate: exp.DocumentReference<T>) {}
+
+  readonly id = this._delegate.id;
+  readonly firestore = new FirebaseFirestore(this._delegate.firestore);
+  readonly path = this._delegate.path;
+
+  get parent(): legacy.CollectionReference<T> {
+    return new CollectionReference<T>(parent(this._delegate));
+  }
+
+  collection(
+    collectionPath: string
+  ): legacy.CollectionReference<legacy.DocumentData> {
+    return new CollectionReference(collection(this._delegate, collectionPath));
+  }
+
+  isEqual(other: DocumentReference<T>): boolean {
+    return refEqual(this._delegate, other._delegate);
+  }
+
+  set(data: Partial<T>, options?: legacy.SetOptions): Promise<void> {
+    if (options) {
+      return setDoc(this._delegate, unwrap(data), options);
+    } else {
+      return setDoc(this._delegate, unwrap(data));
+    }
+  }
+
+  update(data: legacy.UpdateData): Promise<void>;
+  update(
+    field: string | FieldPath,
+    value: any,
+    ...moreFieldsAndValues: any[]
+  ): Promise<void>;
+  update(
+    dataOrField: any,
+    value?: any,
+    ...moreFieldsAndValues: any[]
+  ): Promise<void> {
+    if (arguments.length === 1) {
+      return updateDoc(this._delegate, unwrap(dataOrField));
+    } else {
+      return updateDoc(
+        this._delegate,
+        unwrap(dataOrField),
+        unwrap(value),
+        ...unwrap(moreFieldsAndValues)
+      );
+    }
+  }
+
+  delete(): Promise<void> {
+    return deleteDoc(this._delegate);
+  }
+
+  get(options?: legacy.GetOptions): Promise<DocumentSnapshot<T>> {
+    let snap: Promise<exp.DocumentSnapshot<T>>;
+    if (options?.source === 'cache') {
+      snap = getDocFromCache(this._delegate);
+    } else if (options?.source === 'server') {
+      snap = getDocFromServer(this._delegate);
+    } else {
+      snap = getDoc(this._delegate);
+    }
+    return snap.then(result => new DocumentSnapshot(result));
+  }
+
+  onSnapshot(observer: {
+    next?: (snapshot: DocumentSnapshot<T>) => void;
+    error?: (error: legacy.FirestoreError) => void;
+    complete?: () => void;
+  }): () => void;
+  onSnapshot(
+    options: legacy.SnapshotListenOptions,
+    observer: {
+      next?: (snapshot: DocumentSnapshot<T>) => void;
+      error?: (error: Error) => void;
+      complete?: () => void;
+    }
+  ): () => void;
+  onSnapshot(
+    onNext: (snapshot: DocumentSnapshot<T>) => void,
+    onError?: (error: Error) => void,
+    onCompletion?: () => void
+  ): () => void;
+  onSnapshot(
+    options: legacy.SnapshotListenOptions,
+    onNext: (snapshot: DocumentSnapshot<T>) => void,
+    onError?: (error: Error) => void,
+    onCompletion?: () => void
+  ): () => void;
+  onSnapshot(...args: any): () => void {
+    let options: legacy.SnapshotListenOptions = {};
+    let userObserver: PartialObserver<DocumentSnapshot<T>>;
+
+    if (isPartialObserver(args[0])) {
```

Comment:
optional: There seems to be some code overlap between here and L547. You could probably extract the logic into a shared method, but might not be worth it considering it's just shim testing code.

Reply:
Good idea! Done.